docs/source/proposals/jit-classes.rst

```diff
@@ -0,0 +1,220 @@
+Type information is necessary to materailize the instance (e.g. allocate the
+storage).  The simplest way is to let user provide the type of each attributes
+as well as the ordering; for instance::
+
+    dct = OrderedDict()
+    dct['x'] = int32
+    dct['y'] = float32
+
+Allowing user to supply an ordered dictionary will provide the name, ordering
+and types of the attributes.  However, this statically typed semantic is not as
+flexible as the Python semantic which behaves like a generic class.
+
+Inferring the type of attributes is difficult.  In a previous attempt to
+implement JIT classes, the ``__init__`` method is specialized to capture
+the type stored into the attributes.  Since the method can contain arbitrary
+logic, the problem can become a dependent typing problem if types are assigned
+conditionally depending on the value. (Very few languages implement dependent
+typing and those that does are mostly theorem provers.)
+
+Alternatively, we can let user provide the typing logic.  For example, user
+can supply a function that runs in the type domain::
+
+    def get_type(x, y):
+        dct = OrderedDict()
+        if x == int32:
+            dct['x'] = float32
+        else:
+            raise TypeError
+        dct['y'] = y
+        return dct
+
+    class Foo:
+        def __init__(self, x, y):
+            self.a = x
+            self.b = y
+
+The ``get_type`` takes the same arguments as the ``__init__`` of Foo but in
+the type domain (e.g. ``get_type(typeof(x), typeof(y)``).  In the function,
+the type of each attributes is decided.  The ``get_type`` function requires
+``x`` to be of ``int32`` and uses ``float32`` for attribute ``a``.  Attribute
+``b`` can be of any type and it is the same as the type of ``y``.  This scheme
+will support generic classes without the need of solving a difficult type
+inference problem.
+
+
+Immutable class: jit-struct
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Jit-classes are flexible and they can be too flexible.  The flexibility
+requires runtime assistance that is not available or inefficient in more
+some hardware, such as the GPGPU targets.  One way to limit the flexibility
+is to enforce immutability. We call this a *jit-struct*.  Python has immutable
+containers such as the tuple.  A jit-struct will behave like a jit-class with
+a restriction that disallow mutating any attributes after instantiation.
+In other words, mutation of attributes is only allowed inside ``__init__``.
+
+This can be easily done by changing the type of ``self`` argument passed to
+the ``__init__``.  A jit-struct instance is a pass-by-value structure with
+the exception during instantiation that a pass-by-reference structure is passed
+to ``__init__``.
+
+User API
+~~~~~~~~
+
+
+Typing function as external function
+------------------------------------
+
+.. code-block:: python
+
+    def spec(x, y):
+        dct = OrderedDict()
+        dct['x'] = x
+        dct['y'] = y
+        return dct
+
+    @jit(spec, immutable=False)
```

Comment:
The fact that we can't use the obvious `jit.class` makes it a bit suboptimal, IMO ;-)